pkg/services/ngalert/state/historian/annotation.go

```diff
@@ -30,45 +30,79 @@ func NewAnnotationHistorian(annotations annotations.Repository, dashboards dashb
 	}
 }
 
-func (h *AnnotationStateHistorian) RecordState(ctx context.Context, rule *ngmodels.AlertRule, labels data.Labels, evaluatedAt time.Time, currentData, previousData state.InstanceStateAndReason) {
+// RecordStates writes a number of state transitions for a given rule to state history.
+func (h *AnnotationStateHistorian) RecordStates(ctx context.Context, rule *ngmodels.AlertRule, states []state.StateTransition) {
 	logger := h.log.New(rule.GetKey().LogContext()...)
```

Comment:
nit (no need to address if you do not change anything else): we use context for propagating rule key added in https://github.com/grafana/grafana/pull/57476. You can just do `logger := h.log.FromContext(ctx)`